### PR TITLE
HPCC-27871 Track k8s resources created by processes and ensure cleanup

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by richardkchapman — CHANGES_REQUESTED** [review]

File: common/workunit/workunit.cpp [hunk 1]
```diff
@@ -14489,33 +14496,40 @@ bool applyK8sYaml(const char *componentName, const char *wuid, const char *job,
 #endif
 
     runKubectlCommand(componentName, "kubectl replace --force -f -", jobYaml, nullptr);
+
+    if (autoCleanup)
+    {
+        // touch a file, with naming convention { componentName }-{ resourceType }-{ jobName }.k8s
+        // it will be used if the job fails ungracefully, to tidy up leaked resources
+        // normally (during graceful cleanup) these resources and files will be deleted by deleteK8sResource
+        VStringBuffer k8sResourcesFilename("%s,%s,%s.k8s", componentName, resourceType, jobName.str());
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/workunit/workunit.cpp:14505]: I'll fix comment, changed from '-' to ',' because '-' is used in workunit id's, and therefore makes it hard to parse (in clean up script)

[the remark this answers is not in the document]
Comment (jakesmith) [reply common/workunit/workunit.cpp:14505]:
```
  IFS=, read componentName resourceType jobName <<< $(basename ${filename} .k8s)
  resourceName="${componentName}-${resourceType}-${jobName}"
```
The cleanup script parses as ',' separated, in order to extract the componentName, resourceType and jobName, then builds up resourceName, and uses resourceType and resourceName to delete the resource.

File: system/jlib/jfile.cpp [hunk 2]
```diff
@@ -4301,6 +4301,11 @@ IFile * createIFile(const char * filename)
 #endif
 }
 
+void touchFile(const char *filename)
+{
+    Owned<IFile> iFile = createIFile(filename);
```
Comment (richardkchapman) [inline system/jlib/jfile.cpp:4306]: Any possibility that this will fail, and if so, should you handle exceptions?

**Review by richardkchapman — DISMISSED** [review]
One question

File: system/jlib/jfile.cpp (hunk 2 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:4306]: I think correct for touchFile to throw an exception, and ok (if unlikely) for createIFile to throw one on its behalf.
but the IFile->open could in theory return null.
I'll add a check for that and throw an exception if it does.

Comment (jakesmith) [reply system/jlib/jfile.cpp:4306]: @richardkchapman - please see pushed change

File: common/workunit/workunit.cpp [hunk 3]
```diff
@@ -14458,12 +14465,12 @@ bool applyK8sYaml(const char *componentName, const char *wuid, const char *job,
...
         return false;
     }
-    jobYaml.replaceString("_HPCC_JOBNAME_", jobname.str());
+    jobYaml.replaceString("_HPCC_JOBNAME_", jobName.str());
 
     VStringBuffer args("\"--workunit=%s\"", wuid);
     for (const auto &p: extraParams)
     {
-        if (hasPrefix(p.first.c_str(), "_HPCC_", false)) // jobspec substituion
+        if (hasPrefix(p.first.c_str(), "_HPCC_", false)) // job yaml substituion
```
Comment (ghalliday) [inline common/workunit/workunit.cpp:14473]: trivial typo: substitu**t**ion

**Review by ghalliday — APPROVED** [review]

File: common/workunit/workunit.cpp (hunk 3 above)
Comment (jakesmith) [reply common/workunit/workunit.cpp]: @ghalliday - fixed

**Review by ghalliday — APPROVED** [review]

**Comment by richardkchapman** [comment]
@jakesmith Looks good, please squash

**Review by richardkchapman — APPROVED** [review]

**Comment by jakesmith** [comment]
@richardkchapman - squashed
